content/src/ports/denylist.ts

```diff
@@ -1,50 +1,104 @@
-  const bannedList = new Set()
-
+export async function createDenylist(
+  components: Pick<AppComponents, 'env' | 'logs' | 'fetcher'> & {
+    fs: Pick<FSComponent, 'createReadStream' | 'existPath'>
+  }
+): Promise<Denylist & IBaseComponent> {
+  const logger = components.logs.getLogger('Denylist')
+  const deniedContentIdentifiers = new Set()
   const fileName = resolve(
     components.env.getConfig(EnvironmentConfig.STORAGE_ROOT_FOLDER),
     components.env.getConfig(EnvironmentConfig.DENYLIST_FILE_NAME)
   )
 
-  try {
-    await components.fs.existPath(fileName)
-
-    const content = components.fs.createReadStream(fileName, {
-      encoding: 'utf-8'
-    })
-
-    const lines = createInterface({
-      input: content,
-      crlfDelay: Infinity
+  const denylistsUrlsRaw = components.env.getConfig<string>(EnvironmentConfig.DENYLIST_URLS)?.split(/[\r\n\s]+/gm) ?? []
+  const denylistsUrls: URL[] = denylistsUrlsRaw
+    .filter((url) => {
+      try {
+        new URL(url)
+        return true
+      } catch {
+        return false
+      }
     })
+    .map((url) => new URL(url))
+  logger.info(`File: ${fileName}`)
+  logger.info(`Urls: ${denylistsUrls}`)
 
+  const processLines = async (lines: Iterable<string> | AsyncIterable<string>) => {
     for await (const line of lines) {
-      bannedList.add((line as string).trim())
+      try {
+        const cid = line.trim()
+        if (!cid.startsWith('#')) {
+          deniedContentIdentifiers.add((line as string).trim())
+        }
+      } catch (err: any) {
+        logger.error(err)
+      }
     }
+  }
+
+  const loadDenylistFromFile = async () => {
+    if (!fileName) return
+    if (!(await components.fs.existPath(fileName))) {
+      logger.info("Denylist file doesn't exist", { fileName })
+      return
+    }
+
+    const content = components.fs.createReadStream(fileName, { encoding: 'utf-8' })
 
-    logs.info('Load successfully the denylist')
-  } catch (err) {
-    console.log('ERR', err)
-    logs.warn("Couldn't load a denylist")
+    const lines = createInterface({ input: content, crlfDelay: Infinity })
+
+    await processLines(lines)
+  }
+
+  const loadDenylistFromUrl = async (downloadableLink: URL) => {
+    try {
+      const response = await components.fetcher.fetch(downloadableLink.toString())
+      const content = await response.text()
+      const lines = content ? content.split(/[\r\n\s]+/) : []
+      await processLines(lines)
+    } catch (err: any) {
+      logger.error(err)
+    }
   }
 
-  const isDenyListed = (hash: string): boolean => {
-    return bannedList.has(hash)
+  const loadDenylists = async () => {
+    try {
+      const originalSize = deniedContentIdentifiers.size
+      await loadDenylistFromFile()
+
+      for (const url of denylistsUrls) {
+        await loadDenylistFromUrl(url)
+      }
+      if (deniedContentIdentifiers.size > originalSize) {
+        logger.info('New elements where added')
```

Comment:
Can it happen that the elements changed but the size didn't?

Reply:
Yes, you're right... didn't want to log every 2 minutes in each reload as it would overflood the logs.. will delete it